src/Features/LanguageServer/Protocol/Extensions/Extensions.cs

```diff
@@ -50,6 +51,27 @@ public static ImmutableArray<TextDocument> GetTextDocuments(this Solution soluti
             return documentIds.SelectAsArray(id => getDocument(solution, id));
         }
 
+        public static (Document?, Solution) GetDocumentAndSolution(this ILspSolutionProvider provider, TextDocumentIdentifier? textDocument, string? clientName)
```

Comment:
I wonder if this should just be an ILspSolutionProvider method instead of an extension

Reply:
Good idea. I think originally I was hoping to remove ILspSolutionProvider altogether, but that can't happen until the TypeScript shims are removed anyway, and when that happens this method can move entirely into the RequestExecutionQueue altogether regardless.

Reply:
Upon looking at this again, I actually think this belongs here. All of these extension methods just pull data from solution provider, just like this one. Adding this to solution provider direct would end up basically having identical implementations in both concretes, so I think an extension fits.